Add support for Topology Manager to Device Manager

klueska opened the pull request:
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:
This PR adds the functionality required to turn the `devicemanager` into a `TopologyManager` `HintProvider`.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

Commit:
Added stub support for Topology Manager to Device Manager

Co-authored-by: Conor Nolan <conor.nolan@intel.com>
Co-authored-by: Sreemanti Ghosh <sreemanti.ghosh@intel.com>
Co-authored-by: Kevin Klues <kklues@nvidia.com>

klueska commented:
/test pull-kubernetes-integration

igsilya commented on pkg/kubelet/cm/devicemanager/topology_hints.go:
```diff
@@ -0,0 +1,139 @@
+
+	// Iterate through all combinations of devices and build hints from them.
+	hints := []topologymanager.TopologyHint{}
+	m.iterateDeviceCombinations(devices, amount, func(combination sets.String) {
```
This looks awfully slow. IIUC, computational complexity for this loop is O(amount *  nDevices! / (amount! * (nDevices - amount)!)).

Modern SRIOV capable NICs usually has 64+ VFs. So, if someone will try to allocate 32 VFs for a single container this loop will try to generate over 10^18 hints and will hang here forever until the memory exhausted.
Maybe it's better to group all devices into buckets by socket_id and only iterate over socket combinations similar to what cpu manager does?
Number of sockets also could be too big, but it's fairly low in most of modern platforms. Hope we could handle about 10^4 hints without significant issues. However, this, IMHO, should be reworked in the future too.

klueska replied:
Yes. You are right.

I have an alternative that I implemented for the CPUManager that is O(num_cpus * num_socket_combinations) which should be OK since the number of sockets tends to be low: #73920

I plan to port that over here before it gets merged.

igsilya replied:
Thanks.

igsilya replied:
Thanks!
One more issue here that also might be covered by switching to sockets combination is that we have unordered list of devices. That means that for mask (1,1) we could allocate few devices from node0 and a few devices from node1 even if we able to allocate all of them from the same node. Using socket based combinations we'll have devices actually sorted by the socket_id and will be able to allocate them as local to each other as possible.

klueska replied:
Order shouldn’t matter here b cause this is just about the hint generation to figure out what affinity different combinations of devices have. Tja allocation happens at the GetAffinity() call where the combined hint from multiple components is consulted and devices are allocated with alignment to the socket mask int the combined hint.

Either way though, the socket combination enumeration is the right strategy no matter what because we should have one possible hint for each socket combination and need to check it no matter what.

igsilya replied:
> Order shouldn’t matter here b cause this is just about the hint generation to figure out what affinity different combinations of devices have. Tja allocation happens at the GetAffinity() call where the combined hint from multiple components is consulted and devices are allocated with alignment to the socket mask int the combined hint.

Yes, you're right that order doesn't matter here while generating hints, but it matters in devicesToAllocate() function which has almost same code part. And if the resulted topology hint will have more than one node, allocated devices could be randomly distributed among them due to unsorted nature of the device list.

> Either way though, the socket combination enumeration is the right strategy no matter what because we should have one possible hint for each socket combination and need to check it no matter what.

Sure.

klueska replied:
Makes sense. I’ll need to double check what I’m doing on the allocation side and I’ll fix it up appropriately.

klueska committed:
Generate TopologyHints from the devicemanager

klueska committed:
Add devicemanager tests for TopologyHint generation

klueska committed:
Consume TopologyHints in the devicemanager

klueska committed:
Add devicemanager tests for TopologyHint consumption

klueska commented:
/test pull-kubernetes-kubemark-e2e-gce-big

klueska commented:
/retest

cbf123 commented:
Do we document anywhere what the expectation is for device plugins to report in the case where the kernel is NUMA-unaware, or where the kernel is NUMA-aware but devices are not NUMA-affined?  (Like in a virtual-machine where the "numa_node" returns a value of "-1".)

The original spec at https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0035-20190130-topology-manager.md#changes-to-existing-components showed a default socketID of "-1", but the code in this PR seems to not be checking for that case.

The context here is the discussion at https://github.com/intel/sriov-network-device-plugin/pull/160 around how to handle the case where the "numa_node" file for the device can't be read, or returns -1.

klueska commented:
The current assumption is that you simply wouldn’t set the field if you couldn’t determine alignment. This should be added to the KEP and the documentation to make that clear.

klueska commented on pkg/kubelet/cm/devicemanager/topology_hints.go:
```diff
@@ -0,0 +1,128 @@
+	return deviceHints
+}
+
+func (m *ManagerImpl) deviceHasTopologyAlignment(resource string) bool {
```
A discussion is going on here https://github.com/intel/sriov-network-device-plugin/pull/160#discussion_r316146230 about whether these are the right semantics or not.

klueska commented:
/assign @ConnorDoyle

ConnorDoyle commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -633,7 +650,72 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
+			socket := -1
+			if m.allDevices[resource][d].Topology != nil &&
+				m.allDevices[resource][d].Topology.Node != nil {
+				socket = int(m.allDevices[resource][d].Topology.Node.ID)
```
This isn't a safe assumption to make. Some architectures support multiple numa nodes per socket. It appears the implementation of the device topology info diverged a bit from the KEP, which specified SocketID as a topology key. (see https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0035-20190130-topology-manager.md#changes-to-existing-components)

Since the topology info contains NUMA node instead of socket ID, here we need to map from NUMA node to socketID. Otherwise we're comparing apples and oranges.

klueska replied:
I agree in general. But when discussing this in the PR review for the device topology info, we looked into what the CPUManager is doing and it does a 1:1 mapping of the NUMA Node returned by cadvisor to set its SocketID:

https://github.com/kubernetes/kubernetes/pull/74423#discussion_r304548485

i.e.
This code in the CPUManager walks `machineInfo.Topology`:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/topology/topology.go#L150

Which is made up of a slice of NUMA nodes:
https://godoc.org/github.com/google/cadvisor/info/v1#MachineInfo

Which, in turn, blurs the line between a NUMA node and a socket by bundling together cores, cache, and memory into a single entity:
https://godoc.org/github.com/google/cadvisor/info/v1#Node

So while I agree that (conceptually) these are apples and oranges, I don't see an immediate way to do the NUMA -> Socket ID mapping you mention.

In fact, it seems that since we pull the list of "sockets" directly from `machineInfo`, we actually are comparing the field in the device Topology to the NUMA ID and not the socketID (even though we call it socketID in the code).

igsilya replied:
To be fair, the first one who compares apples and oranges is cadvisor that thinks that 'package id' in the output of /proc/cpuinfo is the NUMA node id. Moreover it uses correct  /sys/bus/cpu/devices/cpu*/nodeX value for aarch64. So, on armv8 Topology.Node.Id will represent NUMA node, but for all other platforms there will be CPU socket id, because it is 'package id' field of /proc/cpuinfo:
https://github.com/google/cadvisor/blob/75252bdc913a94edbf1eaab517b7a7e51999c898/machine/machine.go#L267

Another thing to think about is that NUMA->SocketID is not the many-to-one relation. It's many-to-many, because there might be several CPU sockets on a single NUMA too. It's easy to make such topology in a VM.

BTW, by default Nova uses 'sockets' while configuring VM CPU topology, i.e. if you'll ask for 10 CPUs with threads policy "preferred" in OpenStack, Nova will create VM with 5 sockets, 1 core on each socket, 2 threads on each core. At the same time you will have no NUMA topology for devices (but it's possible to configure it by hands in libvirt XML).

ConnorDoyle replied:
Thanks @igsilya for the extra info about cadvisor.

To summarize:
* cadvisor uses package ID to fill in NUMA node ID (except on arm)
* cpu manager interprets NUMA node ID from cadvisor as socket ID
* topology manager compares these IDs to device plugin supplied NUMA node ID

Besides some misleading naming, my main question right now is whether we’ll run into problems if the device plugin uses libnuma directly on a machine where top level NUMA nodes don’t correspond 1:1 with CPU package IDs.

For topology manager alpha, we should try to explain this as errata in the usage docs, and try to make all of these more consistent for beta.

/cc @derekwaynecarr

kad replied:
And please, please, don't repeat the same mistake again with NUMA field when communicating between components (e.g. from device plugins towards kubelet). We must have a data type that be able to represent:
- N/A (meaning that device is not aligned to any of NUMA boundaries and it shouldn't be counted in the decision)
- single value (strict memory controller boundary)
- multiple memory controller nodes

Don't reinvent wheel. Use please something in the interfaces that later can be utilized to set in `cpuset/cpuset.mems`

klueska replied:
I did a quick test in a VM that did not have a 1:1 mapping between NUMA ID and Socket ID and it confused the TopologyManager in terms of guaranteeing alignment between the `CPUManager` and the `devicemanager`.

Then adding this (obviously unusable) patch to the `CPUManager` fixed it:
```
diff --git a/pkg/kubelet/cm/cpumanager/topology/topology.go b/pkg/kubelet/cm/cpumanager/topology/topology.go
index f19acac5fc..cd8784d998 100644
--- a/pkg/kubelet/cm/cpumanager/topology/topology.go
+++ b/pkg/kubelet/cm/cpumanager/topology/topology.go
@@ -17,7 +17,13 @@ limitations under the License.
 package topology

 import (
+   "encoding/hex"
    "fmt"
+   "io/ioutil"
+   "path/filepath"
+   "regexp"
+   "strconv"
+   "strings"

    cadvisorapi "github.com/google/cadvisor/info/v1"
    "k8s.io/klog"
@@ -165,6 +171,11 @@ func Discover(machineInfo *cadvisorapi.MachineInfo) (*CPUTopology, error) {
        }
    }

+   err := patchNUMAInfo(CPUDetails)
+   if err != nil {
+       return nil, fmt.Errorf("unable to patch NUMA info for the CPU Topology: %v", err)
+   }
+
    return &CPUTopology{
        NumCPUs:    machineInfo.NumCores,
        NumSockets: len(machineInfo.Topology),
@@ -194,3 +205,68 @@ func getUniqueCoreID(threads []int) (coreID int, err error) {

    return min, nil
 }
+
+// patchNUMAInfo patches the CPUDetails maps with Socket IDs populated as NUMA
+// node IDs from sysfs instead of physical_ids from /proc/cpuinfo.
+func patchNUMAInfo(cpudetails CPUDetails) error {
+   // Get paths to NUMA node directories in sysfs.
+   nodepaths, err := filepath.Glob("/sys/devices/system/node/node*")
+   if err != nil {
+       return err
+   }
+
+   // For each NUMA node directory ...
+   re := regexp.MustCompile("[0-9]+")
+   for _, path := range nodepaths {
+       // Find the nodeID of the NUMA node.
+       nodeID, err := strconv.Atoi(re.FindString(path))
+       if err != nil {
+           return err
+       }
+
+       // Read the 'cpumap' of the NUMA node from sysfs.
+       path := fmt.Sprintf("/sys/devices/system/node/node%d/cpumap", nodeID)
+       cpumap, err := ioutil.ReadFile(path)
+       if err != nil {
+           return err
+       }
+
+       // Convert the 'cpumap' into a bitmask of CPUs in a byte array.
+       _cpumap := strings.TrimSpace(string(cpumap))
+       _cpumap = strings.Replace(_cpumap, ",", "", -1)
+       bitmasks, err := hex.DecodeString(_cpumap)
+       if err != nil {
+           return err
+       }
+
+       // Walk the CPU bitmask byte array and fill an []int slice with CPU
+       // indices. The bitmasks is read from right --> left across the bytes.
+       var cpus []int
+       for i := 0; i < len(bitmasks); i++ {
+           ri := len(bitmasks) - 1 - i
+           for j := 0; j < 8; j++ {
+               if bitmasks[ri]&(1<<byte(j)) > 0 {
+                   cpus = append(cpus, i*8+j)
+               }
+           }
+       }
+
+       // For every cpu in the cpus []int slice, update its SocketID entry in
+       // CPUDetails with the id of this NUMA node (as necessary).
+       for _, cpu := range cpus {
+           if nodeID != cpudetails[cpu].SocketID {
+               klog.V(3).Infof("[cpumanager] Patching SocketID for CPU %v, %v -> %v",
+                   cpu, cpudetails[cpu].SocketID, nodeID)
+               cpudetails[cpu] = CPUInfo{
+                   CoreID:   cpudetails[cpu].CoreID,
+                   SocketID: nodeID,
+               }
+           }
+       }
+   }
+
+   return nil
+}
```

klueska replied:
Because of the weirdness of cadvisor, it looks like the CPUManager does in fact build its topology info as intended (except on arm):

i.e. mapping
SocketID -> CoreID -> CPUID
not
NUMAID -> CoreID -> CPUID

However, the natural association of a device (and the original intended purpose of the TopologyManager) was to provide alignment on a NUMA ID, not a Socket ID.

As such, here is my proposal:
1) Rename the `TopologyHint` field from `SocketAffinity` to `NUMAffinity`
2) Update the hint generation code in the `CPUManager` to set `NUMAAffinity` appropriately by adding logic to discover the mapping between socket ID and NUMA ID

This is kind of the opposite of what @ConnorDoyle proposed (i.e. doing a map of NUMA ID -> Socket ID in the devicemanager). However, I think updating the CPUManager makes more sense, because the alignment we actually want to satisfy in the end is NUMA alignment (not socket alignment).

This proposal will leave the core logic of the CPUManager unchanged, and only update the hint generation to be more in line with what we want in terms of aligning on NUMA. Moreover, the logic in this PR will remain exactly as is (except that we will need to patch it with the new name `NUMAffinity` for the `TopologyHint` field, rather than `SocketAffinity`).

Thoughts?

lmdaly replied:
It sounds to me like we should be updating cAdvisor to have two fields, one which is the NUMA Node and one which is the Socket, because going forward any components that care about Topology Manager may need to distinguish. 

The CPU Manager can continue to build its topology as is except there is not an implicit mapping from NUMA Node field in cAdvisor which is actually the Socket ID.

Then when the hint generation is happening for CPU Manager it will use the NUMA Node information which is now part of cAdvisor to generate hints. 

We add as part of the docs that Device Plugin authors should send back NUMA Node information not Socket ID when writing their plugins. 

And +1 on the name change for Socket Affinity if the above route is chosen, except my suggestion is NUMANodeAffinity to be as explicit as possible.

klueska replied:
I agree in the long term. In the short term, I was proposing we add the logic to do the SocketID -> NUMA ID mapping in the CPUManager (instead of cadvisor) with a long term plan of moving it out to cadvisor.

klueska committed:
Change Topology.NUMANode in device plugin interface to a repeated field

klueska committed:
Update semantics to set Preferred field in TopologyHint generation

We now only set Preferred to true if resources can be allocated with a
size equal to the minimimum _possible_ mask when all resources are
available.

klueska commented:
/priority important-soon
/milestone v1.16

klueska committed:
Update CPUManager topology helpers to accept multiple ids

derekwaynecarr commented:
/milestone v1.16

ConnorDoyle commented on pkg/kubelet/cm/cpumanager/topology_hints_test.go:
```diff
@@ -139,13 +131,38 @@ func TestGetTopologyHints(t *testing.T) {
 			expectedHints: nil,
 		},
+		{
+			name:          "Request 2 CPUs, 1 available on NUMA 0, 1 available on NUMA 1",
```
+1

ConnorDoyle reviewed (approved):
Thanks for adding a new test case for the socket constraints in the CPU manager’s hint generation logic.

/approve
/lgtm

klueska commented:
/test pull-kubernetes-bazel-build

derekwaynecarr commented:
this PR includes the commit from #82020 that updates the device plugin which I previously approved.  in order to mitigate batching of prs in the queue causing rebases, this is fine.

/approve
/lgtm

klueska commented:
/retest
